File: src/familiar_connect/stt/factory.py
"""STT backend selector — dispatches on ``[providers.stt].backend``.

Each per-backend factory takes its typed ``[providers.stt.<backend>]``
sub-table; only secrets (e.g. ``DEEPGRAM_API_KEY``) come from env.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from familiar_connect.stt.deepgram import (
    DeepgramTranscriber,
    create_deepgram_transcriber,
)

if TYPE_CHECKING:
    from familiar_connect.config import STTConfig
    from familiar_connect.stt.protocol import Transcriber


_KNOWN_BACKENDS: frozenset[str] = frozenset({"deepgram", "parakeet", "faster_whisper"})


def create_transcriber(config: STTConfig) -> Transcriber:
    """Build a :class:`Transcriber` from *config*.

    :raises ValueError: backend unknown, or the selected backend's
        required secret (e.g. ``DEEPGRAM_API_KEY``) is missing, or its
        optional extra is not installed. Caller is expected to log +
        degrade — see ``commands/run.py``.
    """
    backend = config.backend
    if backend not in _KNOWN_BACKENDS:
        msg = (
            f"Unknown STT backend {backend!r}. "
            f"Known: {sorted(_KNOWN_BACKENDS)}. "
            f"Set [providers.stt].backend in character.toml."
        )
        raise ValueError(msg)

    if backend == "deepgram":
        return _create_deepgram(config)
    if backend == "parakeet":
        return _create_parakeet(config)
    if backend == "faster_whisper":
        return _create_faster_whisper(config)

    # unreachable while every _KNOWN_BACKENDS entry has a dispatch arm
    msg = f"backend {backend!r} accepted but not dispatched"  # pragma: no cover
    raise ValueError(msg)  # pragma: no cover


def _create_deepgram(config: STTConfig) -> DeepgramTranscriber:
    return create_deepgram_transcriber(config.deepgram)


def _create_parakeet(config: STTConfig) -> Transcriber:
    """Lazy-import Parakeet so the heavy numpy/NeMo deps stay optional."""
    try:
        from familiar_connect.stt.parakeet import (  # noqa: PLC0415
            create_parakeet_transcriber,
        )
    except RuntimeError as exc:
        # numpy missing → ``ParakeetTranscriber`` module raises on import.
        # Re-raise as ValueError so ``run.py`` catches + warns uniformly.
        raise ValueError(str(exc)) from exc
    return create_parakeet_transcriber(config.parakeet)


def _create_faster_whisper(config: STTConfig) -> Transcriber:
    """Lazy-import FasterWhisper so its CT2 deps stay optional."""
    try:
        from familiar_connect.stt.faster_whisper import (  # noqa: PLC0415
            create_faster_whisper_transcriber,
        )
    except RuntimeError as exc:
        # numpy missing → module raises on import; surface as ValueError.
        raise ValueError(str(exc)) from exc
    return create_faster_whisper_transcriber(config.faster_whisper)

File: rust/familiar-connect/src/stt/factory.rs
//! STT backend selector — dispatches on `[providers.stt].backend`
//! (subsystem 09; Python `stt/factory.py`).
//!
//! Each per-backend builder takes its typed `[providers.stt.<backend>]`
//! sub-table; only secrets (`DEEPGRAM_API_KEY`) come from the environment.

use crate::config::STTConfig;
use crate::stt::deepgram::build_deepgram_transcriber;
use crate::stt::protocol::{SttError, Transcriber};

/// Known backends, sorted (the factory error renders this list).
pub const KNOWN_BACKENDS: [&str; 3] = ["deepgram", "faster_whisper", "parakeet"];

/// Build a [`Transcriber`] from `config`, reading `DEEPGRAM_API_KEY` from the
/// environment for the Deepgram backend.
///
/// # Errors
/// - [`SttError::UnknownBackend`] when `backend` is not one of [`KNOWN_BACKENDS`].
/// - [`SttError::MissingApiKey`] when the Deepgram backend is selected but
///   `DEEPGRAM_API_KEY` is unset.
/// - [`SttError::LocalSttUnavailable`] when a local backend (`parakeet` /
///   `faster_whisper`) is selected — no local engine is built into this binary
///   yet (DESIGN §6). The caller logs + degrades to text-only, mirroring the
///   Python lazy-import degradation.
pub fn create_transcriber(config: &STTConfig) -> Result<Box<dyn Transcriber>, SttError> {
    let deepgram_key = std::env::var("DEEPGRAM_API_KEY").ok();
    create_transcriber_with_secrets(config, deepgram_key)
}

/// [`create_transcriber`] with the Deepgram secret injected (hermetic seam: the
/// crate forbids `unsafe`, so tests cannot mutate the process environment).
pub(crate) fn create_transcriber_with_secrets(
    config: &STTConfig,
    deepgram_key: Option<String>,
) -> Result<Box<dyn Transcriber>, SttError> {
    let backend = config.backend.as_str();
    if !KNOWN_BACKENDS.contains(&backend) {
        return Err(SttError::UnknownBackend {
            backend: backend.to_string(),
            known: KNOWN_BACKENDS.join(", "),
        });
    }

    match backend {
        "deepgram" => {
            let t = build_deepgram_transcriber(&config.deepgram, deepgram_key)?;
            Ok(Box::new(t))
        }
        // Parakeet / faster-whisper are buffer-and-finalize local backends. No
        // Rust local-STT engine is wired yet (the `local-stt` feature has no
        // engine chosen — DESIGN §6), so selecting them degrades with a clear
        // error, exactly as the Python lazy-import path re-raises its
        // RuntimeError as ValueError for the run command to catch + warn.
        "parakeet" => Err(SttError::LocalSttUnavailable("parakeet".to_string())),
        "faster_whisper" => Err(SttError::LocalSttUnavailable("faster_whisper".to_string())),
        // Unreachable while every KNOWN_BACKENDS entry has a dispatch arm.
        other => Err(SttError::UnknownBackend {
            backend: other.to_string(),
            known: KNOWN_BACKENDS.join(", "),
        }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::STTConfig;

    #[test]
    fn deepgram_backend_dispatches() {
        let cfg = STTConfig {
            backend: "deepgram".to_string(),
            ..STTConfig::default()
        };
        let t = create_transcriber_with_secrets(&cfg, Some("test-key".to_string())).unwrap();
        assert_eq!(t.backend_name(), "deepgram");
        assert!((t.idle_close_s() - 30.0).abs() < f64::EPSILON);
    }

    #[test]
    fn deepgram_missing_key_is_value_error() {
        let cfg = STTConfig {
            backend: "deepgram".to_string(),
            ..STTConfig::default()
        };
        let err = create_transcriber_with_secrets(&cfg, None).err().unwrap();
        assert!(format!("{err}").contains("DEEPGRAM_API_KEY"));
    }

    #[test]
    fn parakeet_backend_degrades() {
        let cfg = STTConfig {
            backend: "parakeet".to_string(),
            ..STTConfig::default()
        };
        let err = create_transcriber_with_secrets(&cfg, None).err().unwrap();
        assert!(matches!(err, SttError::LocalSttUnavailable(ref b) if b == "parakeet"));
        assert!(format!("{err}").contains("parakeet"));
    }

    #[test]
    fn faster_whisper_backend_degrades() {
        let cfg = STTConfig {
            backend: "faster_whisper".to_string(),
            ..STTConfig::default()
        };
        let err = create_transcriber_with_secrets(&cfg, None).err().unwrap();
        assert!(matches!(err, SttError::LocalSttUnavailable(ref b) if b == "faster_whisper"));
        assert!(format!("{err}").contains("faster_whisper"));
    }

    #[test]
    fn unknown_backend_rejected() {
        let cfg = STTConfig {
            backend: "vosk".to_string(),
            ..STTConfig::default()
        };
        let err = create_transcriber_with_secrets(&cfg, Some("test-key".to_string()))
            .err()
            .unwrap();
        assert!(format!("{err}").contains("vosk"));
        assert!(matches!(err, SttError::UnknownBackend { .. }));
    }

    #[test]
    fn known_backends_sorted() {
        let mut sorted = KNOWN_BACKENDS;
        sorted.sort_unstable();
        assert_eq!(KNOWN_BACKENDS, sorted);
    }
}
